src/sqlfluff/rules/L003.py

```diff
@@ -228,15 +232,17 @@ def _process_raw_stack(
             starting_indent_balance = result_buffer[cached_line_count].indent_balance
 
         working_state = _LineSummary(indent_balance=starting_indent_balance)
-        line_no = 1
-        started = line_no == (cached_line_count + 1)
-        for elem in raw_stack:
+
+        line_no = memory.line_no
+        target_line_no = cached_line_count + 1
+        for idx, elem in enumerate(raw_stack[memory.start_process_raw_idx :]):
```

Comment:
Some unrelated, specific tuning of L003 here. Previously, it was starting from the beginning of `raw_stack` every time, which caused O(n^2) performance. Now, it remembers the last line number and position in `raw_stack`, so it can resume scanning on the "current" line. This shows up as a lot fewer executions of these lines in the Python line profiler and a big reduction in "percent of overall function time".